### PR TITLE
Field: fix wikiURL by URL encoding it - and slight refactors

### LINKED ISSUE #10065: Wikipedia link with question mark
### URL

https://www.openstreetmap.org/edit?way=394705975#map=19/44.81767/20.45303

### How to reproduce the issue?

1. Clear the wikipedia tag.
2. Select language "српски / srpski".
3. Start typing "Кафана „?”". Select such value from the drop-down.
4. Click on the button "view on wikipedia.org".
5. You will get the error page in the browser.

### Screenshot(s) or anything else?

Tag "wikipedia=sr:Кафана_„%3F”" works, tag "wikipedia=sr:Кафана „?”" doesn't.
If you save tag "wikipedia=sr:Кафана „?”" problem will stay in the view (not the edit) mode too.

image

### Which deployed environments do you see the issue in?

Released version at openstreetmap.org/edit

### What version numbers does this issue effect?

2.27.3

### Which browsers are you seeing this problem on?

Chrome

### PR DESCRIPTION
**Pull request by hirako2000**
### Description 

closes #10065 by 

-  URL encoding wikiURL tag on change

#### Refacrtor

- Define wikipedia domain URL and scheme for reuse
- Extract toUrlAnchor as its own function for reuse 

### How was this tested?

- Executed the full unit test suite
- Added a unit tests to cover my changes
- ~Would do with complete assertion on the unit test to check the set wikiURL but was unable to expose it to test suite~
- Verified by running the application locally and auto filling various wikipedia URLs containing characters requiring URL encoding, such as `https://en.wikipedia.org/wiki/%3F_(film)` _and followed by some section header as anchor, directly_.

### CODE REVIEW

File: modules/ui/fields/wikipedia.js [hunk 1]
```diff
@@ -306,6 +307,15 @@ export function uiFieldWikipedia(field, context) {
     }
   }
 
+  function updateEncodedWikiUrl(tagLang, tagArticleTitle, anchor) {
+    const underscoredUrlEncodedTitle = encodeURIComponent(tagArticleTitle.replace(/ /g, '_'));
+    const urlAnchored = toUrlAnchor(anchor);
+    _wikiURL = `${scheme}${tagLang}.${domain}/wiki/${underscoredUrlEncodedTitle}${urlAnchored}`;
```
Comment (1ec5) [inline modules/ui/fields/wikipedia.js]: Perhaps this would be more testable if it returns the URL instead of setting `_wikiURL`.

File: modules/ui/fields/wikipedia.js [hunk 2]
```diff
@@ -11,6 +11,8 @@ import { utilGetSetValue, utilNoAuto, utilRebind } from '../../util';
 
 
 export function uiFieldWikipedia(field, context) {
+  const scheme = 'https://';
+  const domain = 'wikipedia.org';
```
Comment (1ec5) [inline modules/ui/fields/wikipedia.js:15]: This is reasonable, though FYI the _languageCode_.wikipedia.org subdomain pattern, `w/index.php` path, and `wiki/` rewrite rule are all specific to Wikipedia and [configurable by each wiki](https://www.mediawiki.org/wiki/Manual:Short_URL).

File: modules/ui/fields/wikipedia.js [hunk 3]
```diff
@@ -306,6 +307,15 @@ export function uiFieldWikipedia(field, context) {
...
 
+  function updateEncodedWikiUrl(tagLang, tagArticleTitle, anchor) {
+    const underscoredUrlEncodedTitle = encodeURIComponent(tagArticleTitle.replace(/ /g, '_'));
+    const urlAnchored = toUrlAnchor(anchor);
+    _wikiURL = `${scheme}${tagLang}.${domain}/wiki/${underscoredUrlEncodedTitle}${urlAnchored}`;
+  }
+
+  function toUrlAnchor(anchor) {
+    return anchor ? '#' + anchor : '';
+  }
```
Comment (1ec5) [inline modules/ui/fields/wikipedia.js]: This ternary is simple enough to inline. While we’re at it, the anchor should also be encoded. You can apply the same encoding to both the page title and the named anchor, replacing spaces with underscores and calling `encodeURIComponent()`. (There are some [nuanced edge cases](https://www.mediawiki.org/wiki/Manual:PAGENAMEE_encoding) that you probably don’t need to spend time worrying about.)

File: modules/ui/fields/wikipedia.js (hunk 1 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: Good idea!

File: modules/ui/fields/wikipedia.js (hunk 2 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js:15]: the lang as subdomain is already being taken care by pre-existing logic. I believe.

Good catch about the specificities. Is this code meant to deal with non wikipedia rewrite rules? I would happily make it wiki agnostic, but it doesn't seem trivial to retrieve the used rewrite rules. Not present in result search metadata, no idea from where to determine it.

File: modules/ui/fields/wikipedia.js (hunk 3 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: sure thing, will refactor. 

reading through the nuanced edge cases, xin cảm ơn.

File: modules/ui/fields/wikipedia.js (hunk 1 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: Made the function return, and assign to wiki so that the tests can invoke it.

File: modules/ui/fields/wikipedia.js (hunk 3 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: I came across the fact URI encoding the whole path (that is including the hash/ref of the anchor) causes an unnecessary redirect. e.g [https://en.wikipedia.org/wiki/%3F_(film)%23Themes_and_style](https://en.wikipedia.org/wiki/%3F_(film)%23Themes_and_style)

I have refactored the function to keep inline use, and dealt with encoding each part, the title and anchor fragment separately.

Note: I'm unable to manually test via the UI assignment of an anchored wiki URI.

File: modules/ui/fields/wikipedia.js (hunk 2 above)
Comment (1ec5) [reply modules/ui/fields/wikipedia.js:15]: You’re right, this feature is 100% specific to Wikipedia. I just pointed it out since the code is becoming a bit more generic in this PR, but I wouldn’t expect any change as a result.

File: modules/ui/fields/wikipedia.js [hunk 4]
```diff
@@ -307,15 +297,25 @@ export function uiFieldWikipedia(field, context) {
...
-  }
+  wiki.anchorFragment = (anchor) => {
+    if (!anchor) {
+      return '';
+    }
+
+    try {
+      // Best-effort `anchorencode:` implementation
+      anchor = anchor.replace(/ /g, '_').replace(/%/g, '.');
+    } catch (e) {
```
Comment (1ec5) [inline modules/ui/fields/wikipedia.js]: Good idea to emulate `anchorencode:`. If it fails for some reason, would it result in an exception that you can catch here, or can we omit the try-catch safely?

File: modules/ui/fields/wikipedia.js [hunk 5]
```diff
@@ -307,15 +297,25 @@ export function uiFieldWikipedia(field, context) {
...
   }
 
-  function updateEncodedWikiUrl(tagLang, tagArticleTitle, anchor) {
-    const underscoredUrlEncodedTitle = encodeURIComponent(tagArticleTitle.replace(/ /g, '_'));
-    const urlAnchored = toUrlAnchor(anchor);
-    _wikiURL = `${scheme}${tagLang}.${domain}/wiki/${underscoredUrlEncodedTitle}${urlAnchored}`;
-  }
+  wiki.encodePath = (tagArticleTitle, anchor) => {
+    const underscoredTitle = tagArticleTitle.replace(/ /g, '_');
+    return `${encodeURIComponent(underscoredTitle)}${anchor ? '#' + encodeURIComponent(wiki.anchorFragment(anchor)): ''}`;
```
Comment (1ec5) [inline modules/ui/fields/wikipedia.js]: Nit: This line is getting a little unwieldy. Inlining `toUrlAnchor()` was good, but we don’t have to put everything on a single line. It’s OK to declare a `let` variable, for instance, and append to it when necessary.

**Review by 1ec5 — COMMENTED** [review]
Thanks for iterating – this is getting close.

File: modules/ui/fields/wikipedia.js (hunk 5 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: agreed, and fixed. 

I didn't make use of a `let` declaration though, for clear variable naming.

File: modules/ui/fields/wikipedia.js (hunk 4 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: I refactored it and yes I believe it could go away, so it is now gone. Please note that _try-catch_ `anchorencode` was in there before this fix. It got extracted into a separate function without  affecting the flow.

I dug further into it, and couldn't find any reason* for replacing `%` with `.`

I concluded that it was a Best-effort at trying to deal with URI encoded looking characters that happen to be contained in sections then made anchors, without resorting to `encodeURIComponent()`, here also for reasons I could not explain. Also, could not think of any possible string content that would make `replace()` throw an error, it may throw error on `null`, `undefined` value, and invalid regex though. But then it would throw an error within the catch block.

Anyhow I have added further unit test cases, which hopefully clarifies intents of those functions now. If the test cases aren't correct, happy to iterate again.


*or was it perhaps an attempt to defend against [Double Encoding Attacks](https://owasp.org/www-community/Double_Encoding) ? If that could be the case perhaps it should be put back where it was, which was the rational of keeping it in there initially, but in any case, it does not seem like a solid way to prevent URL tempering  attacks so I settled with eliminating it.

**Comment by hirako2000** [comment]
> getting close.

Getting close, I don't know. But closer, **yes**. Thanks for your valuable review comments.

File: modules/ui/fields/wikipedia.js (hunk 3 above)
Comment (hirako2000) [reply modules/ui/fields/wikipedia.js]: I was finally able to manually test with anchored URL, figured out how.

File: modules/ui/fields/wikipedia.js (hunk 4 above)
Comment (1ec5) [reply modules/ui/fields/wikipedia.js]: It looks like my recollection of MediaWiki’s behavior may be somewhat outdated. If a section is titled something like “[#time](https://www.mediawiki.org/wiki/Help:Extension:ParserFunctions##time)”, there will be two named anchors, `#time` and `.23time`. Some other reserved characters behave similarly. It used to be the case that the page would contain only `.23time`, but now that it contains `#time` as well, there’s no need for anything special, other than replacing spaces with underscores.

**Review by 1ec5 — APPROVED** [review]

**Comment by tyrasd** [comment]
Right! This should be fixed in iD with https://github.com/openstreetmap/iD/commit/ee56b9c8ae1ec1e5f7544b9644409b82e4e63e3a
